dali/pipeline/operators/detection/box_encoder.cc

```diff
@@ -0,0 +1,157 @@
+// Copyright (c) 2017-2018, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <algorithm>
+
+#include "dali/pipeline/operators/detection/box_encoder.h"
+
+namespace dali {
+void BoxEncoder<CPUBackend>::CalculateIousForBox(float *ious, const BoundingBox &box) const {
+  ious[0] = box.IntersectionOverUnion(anchors_[0]);
+  int best_idx = 0;
+  float best_iou = ious[0];
+
+  for (int anchor_idx = 1; anchor_idx < NumAnchors(); ++anchor_idx) {
+    ious[anchor_idx] = box.IntersectionOverUnion(anchors_[anchor_idx]);
+
+    if (ious[anchor_idx] >= best_iou) {
+      best_iou = ious[anchor_idx];
+      best_idx = anchor_idx;
+    }
+  }
+
+  // For best default box matched with current object let iou = 2, to make sure there is a match,
+  // as this object will be the best (highest IOU), for this default box
+  ious[best_idx] = 2.;
+}
+
+vector<float> BoxEncoder<CPUBackend>::CalculateIous(const vector<BoundingBox> &boxes) const {
+  vector<float> ious(boxes.size() * NumAnchors());
+
+  for (auto bbox_idx = 0u; bbox_idx < boxes.size(); ++bbox_idx) {
+    auto ious_row = ious.data() + bbox_idx * NumAnchors();
+    CalculateIousForBox(ious_row, boxes[bbox_idx]);
+  }
+
+  return ious;
+}
+
+int BoxEncoder<CPUBackend>::FindBestBoxForAnchor(
+  int anchor_idx, const vector<float> &ious, int num_boxes) const {
+  int best_idx = 0;
+  float best_iou = ious[anchor_idx];
+
+  for (int bbox_idx = 1; bbox_idx < num_boxes; ++bbox_idx) {
+    if (ious[bbox_idx * NumAnchors() + anchor_idx] >= best_iou) {
+      best_iou = ious[bbox_idx * NumAnchors() + anchor_idx];
+      best_idx = bbox_idx;
+    }
+  }
+
+  return best_idx;
+}
+
+void BoxEncoder<CPUBackend>::MatchBoxesWithAnchors(
+  const vector<BoundingBox> &boxes, const int *labels, float *out_boxes, int *out_labels) const {
+  int num_boxes = boxes.size();
```

Comment:
could be const. Also consider making all these 'num, count, size' types unsigned

Reply:
Fixed